src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -491,41 +491,46 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
                     this.super_place(place, context, location);
                     match proj.elem {
                         ProjectionElem::Deref => {
-                            this.add(Qualif::NOT_CONST);
-
-                            let base_ty = proj.base.ty(this.mir, this.tcx).to_ty(this.tcx);
-                            if let ty::TyRawPtr(_) = base_ty.sty {
-                                if this.mode != Mode::Fn {
-                                    let mut err = struct_span_err!(
-                                        this.tcx.sess,
-                                        this.span,
-                                        E0396,
-                                        "raw pointers cannot be dereferenced in {}s",
-                                        this.mode
-                                    );
-                                    err.span_label(this.span,
-                                                   "dereference of raw pointer in constant");
-                                    if this.tcx.sess.teach(&err.get_code().unwrap()) {
-                                        err.note(
-                                            "The value behind a raw pointer can't be determined \
-                                             at compile-time (or even link-time), which means it \
-                                             can't be used in a constant expression."
+                            if let Mode::Fn = this.mode {
+                                this.add(Qualif::NOT_CONST);
+                            } else {
+                                let base_ty = proj.base.ty(this.mir, this.tcx).to_ty(this.tcx);
+                                if let ty::TyRawPtr(_) = base_ty.sty {
```

Comment:
> All derefs are NOT_CONST, because we can never prove anything strong enough about the pointer/reference.

How that, safe references should be fine?

Reply:
Oh your comment in https://github.com/rust-lang/rust/issues/54224 indicates this is about promotion. Then never mind, I anyway have no idea what happens there.^^

Reply:
The *type* is safe, but the value isn't guaranteed to be valid / not point to a `static`.

Reply:
I can't parse the end of your sentence, do you mean "isn't guaranteed to point to a static" or "isn't guaranteed to NOT point to a static"?

Reply:
"isn't guaranteed to (be valid | not point to a `static`)"
(as in, pointing to a `static` is as invalid as a random unsafe pointer, *for promotion* - since the value may be different at runtime)

Reply:
Uh, what? I would have thoughts statics are the least problematic to point at in a promoted (aka another static)...?

Reply:
Maybe in a constant, yeah, but in the general case, a static could've changed at runtime.